### ISSUE TITLE
Developer documentation refactor proposal

### ISSUE THREAD

**Issue by petioptrv** [CoinAlpha/hummingbot#4875]
# Tl;dr
- Remove implementation details from documentation; leave only architecture concepts.
- Move all docs to the "gold standard" example connectors and point developers to those (which is probably how exchange implementers already work).
- This will reduce documentation maintenance cost and ensure the docs are up to date.
- Inspired by `pandas`'s approach.

# Why
Currently, the developer docs on implementing connectors is quite detailed and implementation-specific. We offer diagrams dictating even the private variables and methods the different classes should have.

This approach requires us to update the documentation every time we make an important architectural change to the connectors. The fact that the documentation for perpetual connectors was recently updated twice in a couple of months, demonstrates that those changes happen often. This, in turn, adds considerable maintenance overhead while our docs are almost always out of date with the code.

In addition, I personally believe that developers currently rely heavily on specific code examples when implementing exchanges and they have to reconcile what they see in the code with what the documentation is instructing them to do.

# Research
I decided to investigate how other open source projects approach this problem. The `pandas` library offers the possibility to users to implement custom extension data types that can then be used across the `pandas` ecosystem. This task is not as complex as developing a `hummingbot` connector, but from experience, it is not a trivial one either.

Here is [the documentation](https://pandas.pydata.org/docs/development/extending.html) for extending `pandas`, and here is the [`ExtensionDtype`](https://github.com/pandas-dev/pandas/blob/v1.3.4/pandas/core/dtypes/base.py#L35-L368) base class. The documentation is quite sparse and offers the main ideas behind the implementation, while the class itself has extensive docstrings.

# What
I propose that we take a similar approach. In the documentation website, we feature the broad architectural overview: diagrams of how the different classes relate to each-other and a broad explanation of each class' responsibility.

Next, we point the developer to two code examples, one spot and one perpetual connector, and we inform them that those are the "gold standard" connectors that they should base their implementation on. We also inform them that, if they face a particular challenge with their exchange (e.g. the exchange does not offer a given websocket endpoint), they should first post a comment on the relevant channel, and maybe an engineer will point them to a particular exchange that has the same particularity so they can have an example of how that was handled in the codebase.

Finally, we don't need to scrap the documentation we already have regarding the specific functions; those can go into docstrings for those function on the "gold standard" exchanges.

This approach will make it easier for us to maintain the documentation as it will live beside the changes that we make to the connectors; both the code change and the docs update will be in the same PR.

**Comment by dennisocana** [comment]
Thanks Petio, I will have a 1:1 with Mike next week and will add this to our discussion.

**Comment by dennisocana** [comment]
We have agreed to go with this approach and start implementing the changes to both Binance and Binance Perpetual connectors. Petio will create the said tasks.
